master/internal/api_user.go

```diff
@@ -340,6 +350,43 @@ func (a *apiServer) PatchUser(
 		}
 	}
 
+	if req.User.Username != nil {
+		if err = user.AuthZProvider.Get().CanSetUsersUsername(*curUser, targetUser); err != nil {
+			if ok, findErr := user.AuthZProvider.Get().CanGetUser(*curUser, targetUser); err != nil {
+				return nil, findErr
+			} else if !ok {
+				return nil, errUserNotFound
+			}
+			return nil, err
+		}
+
+		// Remove non-ASCII chars to avoid hidden whitespace, confusable letters, etc.
+		username := strings.TrimSpace(latinText.ReplaceAllLiteralString(req.User.Username.Value, ""))
+		// Reject SSO and SCIM
```

Comment:
I recommend simply rejecting invalid usernames rather than fixing them

Reply:
I created a `clearUsername` function for both display name and username. They will throw an error for non-ASCII characters, minimum length, admin/determined conflict, etc.

Reply:
*in the latest commit
thanks for the suggestions